src/rust/ide/lib/view/src/graph_editor/component/breadcrumbs.rs

```diff
@@ -0,0 +1,337 @@
+
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct BreadcrumbsModel {
+    logger                : Logger,
+    animations            : Animations,
+    display_object        : display::object::Instance,
+    pub project_name      : ProjectName,
+    breadcrumbs_container : display::object::Instance,
+    scene                 : Scene,
+    breadcrumbs           : Rc<RefCell<Vec<Breadcrumb>>>,
+    breadcrumb_pop        : frp::Source,
+    breadcrumb_push       : frp::Source<(Rc<MethodPointer>,uuid::Uuid)>,
+    current_index         : Rc<Cell<usize>>
+}
+
+impl BreadcrumbsModel {
+    /// Create new BreadcrumbsModel.
+    pub fn new<'t,S:Into<&'t Scene>>(scene:S, frp:&Frp, focus_manager:&FocusManager) -> Self {
+        let scene                 = scene.into();
+        let project_name          = ProjectName::new(scene,focus_manager);
+        let logger                = Logger::new("Breadcrumbs");
+        let display_object        = display::object::Instance::new(&logger);
+        let breadcrumbs_container = display::object::Instance::new(&logger);
+        let animations            = Animations::new(&frp.network);
+        let scene                 = scene.clone_ref();
+        let breadcrumbs           = default();
+        let breadcrumb_pop        = frp.outputs.breadcrumb_pop.clone_ref();
+        let breadcrumb_push       = frp.outputs.breadcrumb_push.clone_ref();
+        let current_index         = default();
+        Self{logger,display_object,animations,scene,breadcrumbs,breadcrumb_pop,breadcrumb_push,
+            project_name,breadcrumbs_container,current_index}.init()
+    }
+
+    fn init(self) -> Self {
+        self.add_child(&self.project_name);
+        self.add_child(&self.breadcrumbs_container);
+        self.project_name.set_position(Vector3::new(HORIZONTAL_MARGIN,0.0,0.0));
+        self.set_project_name_width(self.project_name.width());
+        //FIXME[dg]: This is a workaround. Main shouldn't be a MethodPointer.
+        let mut method_pointer = enso_protocol::language_server::MethodPointer::default();
+        method_pointer.name    = "Main".to_string();
+        let method_pointer     = Rc::new(method_pointer);
+        self.push_breadcrumb(&method_pointer,&default());
+        self
+    }
+
+    fn width(&self) -> f32 {
+        self.breadcrumbs.borrow().iter().fold(0.0, |acc,breadcrumb| acc + breadcrumb.width())
+    }
+
+    fn set_project_name_width(&self, width:f32) {
+        self.breadcrumbs_container.set_position(Vector3::new(HORIZONTAL_MARGIN + width,0.0,0.0));
+    }
+
+    fn select_breadcrumb(&self, index:usize) {
+        let current_index = self.current_index.get();
+        match index.cmp(&current_index) {
+            cmp::Ordering::Less => {
+                // If we have more crumbs after `index`, we will pop them.
+                let pop_amount    = current_index - index;
+                for _ in 0..pop_amount {
+                    self.breadcrumb_pop.emit(());
```

Comment:
do not emit events from model methods. Instead, return vector of things back to frp network and connect it there. This is just a rule of thumb - never emit things from model if you really, really are not forced to :)